File: gen.py
import os
import sys

def getContents(text, name):
	start = text.find("<" + name + ">")
	end = text.find("</" + name + ">")
	return text[start+len(name)+2:end]

class Lang:
	def __init__(self, src, indexFuncs):
		self.src = src
		self.body = ""
		self.head = getContents(src, "head")
		self.foot = getContents(src, "foot")
		self.dest = getContents(src, "file")

		self.funcs = {}

		funcsDef = getContents(src, "funcs")

		for f in indexFuncs:
			name = f.split(" ")[0]

			self.funcs[name] = getContents(funcsDef, name)

	def func(self, name, varDict):
		result = self.funcs[name]

		for var in varDict.keys():
			result = result.replace("$" + var, varDict[var])

		self.body += (result + "\n")

	def write(self, dest):
		result = self.head + self.body + self.foot
		file = open(dest + "/" + self.dest, "w")
		file.write(result)
		file.close()



settingsFile = open("settings.txt", "r")
settings = settingsFile.read()
settingsFile.close()

dest = getContents(settings, "dest")
srcs = getContents(settings, "langs")
funcsRaw = getContents(settings, "funcs")

idxFuncs = []

for f in funcsRaw.split("\n"):
	if not f == "":
		idxFuncs.append(f)

langs = []

for l in srcs.split("\n"):
	if not l == "":
		file = open(l, "r")
		rawText = file.read()
		file.close()
		langs.append(Lang(rawText, idxFuncs))

idxFuncsDictTemplate = {}

for func in idxFuncs:
	idxFuncsDictTemplate[func.split(" ")[0]] = func.split(" ")

codeFile = open("code.txt", "r")
code = codeFile.read()
codeFile.close()

for line in code.split("\n"):
	if not line == "":
		parts = line.split("#")
		name = parts[0]

		varDict = {}

		for i in range(1, len(parts)):
			varDict[idxFuncsDictTemplate[name.replace(" ", "")][i]] = parts[i]

		for l in langs:
			l.func(name.replace(" ", ""), varDict)

for l in langs:
	l.write(dest)

File: gen.rb
def getContents(txt, name)
	ts = txt.index("<" + name + ">")
	te = txt.index("</" + name + ">")
	st = ts+name.length+2
	ln = te-st
	return txt[st, ln]
end

class Lang

	def initialize(src, indexFuncs)
		@src = src
		@body = ""
		@head = getContents(src, "head")
		@foot = getContents(src, "foot")
		@dest = getContents(src, "file")

		@funcs = Hash.new()

		funcsDef = getContents(src, "funcs")

		for f in indexFuncs
			name = f.split(" ")[0]

			@funcs.store(name, getContents(funcsDef, name))
		end
	end

	def func(name, varDict)
		result = @funcs[name]

		for var in varDict.keys()
			result = result.sub("$" + var, varDict[var])
		end

		@body += (result + "\n")
	end

	def write(dest)
		result = @head + @body + @foot
		file = File.open(dest + "/" + @dest, "w")
		file.write(result)
		file.close()
	end

end

contents = IO.read("settings.txt")

dest = getContents(contents, "dest")
srcs = getContents(contents, "langs")
funcsRaw = getContents(contents, "funcs")

idxFuncs = Array.new()

for f in funcsRaw.split("\n")
	if f != ""
		idxFuncs.push(f)
	end
end

langs = Array.new()

for l in srcs.split("\n")
	if not l == ""
		rawText = IO.read(l)
		langs.push(Lang.new(rawText, idxFuncs))
	end
end

idxFuncsDictTemplate = Hash.new()

for func in idxFuncs
	idxFuncsDictTemplate.store(func.split(" ")[0] ,func.split(" "))
end

codeFile = File.new("code.txt", "r")
code = codeFile.read()
codeFile.close()

for line in code.split("\n")
	if line != ""
		parts = line.split("#")
		name = parts[0]

		varDict = Hash.new()

		i = 1
		while i < parts.length do
			varDict.store(idxFuncsDictTemplate[name.sub(" ", "")][i], parts[i])
			i = i + 1
		end

		for l in langs
			l.func(name.sub(" ", ""), varDict)
		end
	end
end

for l in langs
	l.write(dest)
end